atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/TransactionScopedCacheImpl.java

```diff
@@ -81,28 +80,43 @@ public Map<Cell, byte[]> get(
         return AtlasFutures.getUnchecked(getAsync(tableReference, cells, valueLoader));
     }
 
+    @Override
+    public Map<Cell, byte[]> getWithCachedRef(
+            TableReference tableReference,
+            Set<Cell> cells,
+            Function<CacheLookupResult, ListenableFuture<Map<Cell, byte[]>>> valueLoader) {
+        ensureNotFinalised();
+        return AtlasFutures.getUnchecked(getAsyncWithCachedRef(tableReference, cells, valueLoader));
+    }
+
     @Override
     public ListenableFuture<Map<Cell, byte[]>> getAsync(
             TableReference tableReference,
             Set<Cell> cells,
             Function<Set<Cell>, ListenableFuture<Map<Cell, byte[]>>> valueLoader) {
         ensureNotFinalised();
+        return getAsyncWithCachedRef(
+                tableReference, cells, cacheLookupResult -> valueLoader.apply(cacheLookupResult.missedCells()));
+    }
+
+    @Override
+    public ListenableFuture<Map<Cell, byte[]>> getAsyncWithCachedRef(
+            TableReference tableReference,
+            Set<Cell> cells,
+            Function<CacheLookupResult, ListenableFuture<Map<Cell, byte[]>>> valueLoader) {
+        ensureNotFinalised();
         // Short-cut all the logic below if the table is not watched.
         if (!valueStore.isWatched(tableReference)) {
-            return valueLoader.apply(cells);
+            return valueLoader.apply(CacheLookupResult.of(Map.of(), cells));
         }
 
         CacheLookupResult cacheLookup = cacheLookup(tableReference, cells);
 
-        if (cacheLookup.missedCells().isEmpty()) {
-            return Futures.immediateFuture(filterEmptyValues(cacheLookup.cacheHits()));
-        } else {
-            return Futures.transform(
-                    valueLoader.apply(cacheLookup.missedCells()),
-                    uncachedValues -> processUncachedCells(
-                            tableReference, cacheLookup.cacheHits(), cacheLookup.missedCells(), uncachedValues),
-                    MoreExecutors.directExecutor());
-        }
```

Comment:
Potentially controversial change here: before if we had no missed cells, we'd just return early. But now we delegate to the value loader, so it can decide if too many cells were cached and throw. Not extra cells should be loaded, though, since we pass an empty Set as the missed cells.

Modified the test for the  existing behaviour here: https://github.com/palantir/atlasdb/pull/6655/files#diff-0205f833db5fc2b2032f684954c530b494e176e210ab17d9bdd0713869ef1981R152-R163